source/nomad_service.go

```diff
@@ -0,0 +1,242 @@
+		for _, hostname := range hostnameList {
+			hnEndpoints, err := ns.generateEndpoints(ctx, namespace, serviceName, annotations, hostname, providerSpecific, setIdentifier)
+			if err != nil {
+				return nil, err
+			}
+
+			endpoints = append(endpoints, hnEndpoints...)
+		}
+	}
+
+	return endpoints, nil
+}
+
+func (ns *nomadServiceSource) endpointsFromTemplate(ctx context.Context, namespace string, serviceName string, annotations map[string]string) ([]*endpoint.Endpoint, error) {
+	hostnames, err := ns.execTemplate(ns.fqdnTemplate, nomadServiceMetadata{Namespace: namespace, Name: serviceName})
+	if err != nil {
+		return nil, err
+	}
+
+	providerSpecific, setIdentifier := getProviderSpecificAnnotations(annotations)
+
+	var endpoints []*endpoint.Endpoint
+	for _, hostname := range hostnames {
+		hnEndpoints, err := ns.generateEndpoints(ctx, namespace, serviceName, annotations, hostname, providerSpecific, setIdentifier)
+		if err != nil {
+			return nil, err
+		}
+		endpoints = append(endpoints, hnEndpoints...)
+	}
+
+	return endpoints, nil
+}
+
+type nomadServiceMetadata struct {
+	Namespace string
+	Name      string
+}
+
+func (ns *nomadServiceSource) execTemplate(tmpl *template.Template, obj nomadServiceMetadata) (hostnames []string, err error) {
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, obj); err != nil {
+		return nil, fmt.Errorf("failed to apply template on service %s/%s: %w", obj.Namespace, obj.Name, err)
+	}
+	for _, name := range strings.Split(buf.String(), ",") {
+		name = strings.TrimFunc(name, unicode.IsSpace)
+		name = strings.TrimSuffix(name, ".")
+		hostnames = append(hostnames, name)
+	}
+	return hostnames, nil
+}
+
+func (ns *nomadServiceSource) generateEndpoints(ctx context.Context, namespace string, serviceName string, annotations map[string]string, hostname string, providerSpecific endpoint.ProviderSpecific, setIdentifier string) (endpoints []*endpoint.Endpoint, _ error) {
+	hostname = strings.TrimSuffix(hostname, ".")
+
+	resource := fmt.Sprintf("service/%s/%s", namespace, serviceName)
+
+	ttl := getTTLFromAnnotations(annotations, resource)
+
+	targets := getTargetsFromTargetAnnotation(annotations)
+
+	if len(targets) == 0 {
+		opts := &nomad.QueryOptions{Namespace: namespace}
+		opts = opts.WithContext(ctx)
+
+		svcRegs, _, err := ns.client.Services().Get(serviceName, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		// Collect unique service addresses
+		svcAddrs := make(map[string]struct{})
+		for _, svcReg := range svcRegs {
+			svcAddrs[svcReg.Address] = struct{}{}
+		}
+
+		for addr := range svcAddrs {
+			targets = append(targets, addr)
+		}
+	}
+
+	endpoints = append(endpoints, endpointsForHostname(hostname, targets, ttl, providerSpecific, setIdentifier, resource)...)
+
+	return endpoints, nil
+}
+
+func (sc *nomadServiceSource) setResourceLabel(namespace string, serviceName string, endpoints []*endpoint.Endpoint) {
+	for _, ep := range endpoints {
```

Comment:
label := fmt.Sprintf("service/%s/%s", namespace, serviceName)
for _, ep := range endpoints {
   ep.Labels[endpoint.ResourceLabelKey] = label
}